rust/agents/relayer/src/settings/whitelist.rs

```diff
@@ -0,0 +1,209 @@
+use std::fmt;
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+use std::num::ParseIntError;
+
+use ethers::types::H256;
+use serde::de::{Error, SeqAccess, Visitor};
+use serde::{Deserialize, Deserializer};
+
+use abacus_core::AbacusMessage;
+
+/// Whitelist defining which messages should be relayed. If no wishlist is provided ALL
+/// messages will be relayed.
+///
+/// Valid options for each of the tuple elements are
+/// - wildcard "*"
+/// - single value in decimal or hex (must start with `0x`) format
+/// - list of values in decimal or hex format
+///
+/// 4-tuple in the form `(sourceDomain, sourceAddress, destinationDomain, destinationAddress)`.
+#[derive(Debug, Deserialize, Default, Clone)]
+#[serde(transparent)]
+pub struct Whitelist(Option<Vec<WhitelistElement>>);
+
+#[derive(Debug, Clone)]
+enum Filter<T> {
+    Wildcard,
+    Enumerated(Vec<T>),
+}
+
+impl<T: PartialEq> Filter<T> {
+    fn matches(&self, v: &T) -> bool {
+        match self {
+            Filter::Wildcard => true,
+            Filter::Enumerated(list) => list.iter().any(|i| i == v),
+        }
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum StrOrInt<'a> {
+    Str(&'a str),
+    Int(u32),
+}
+
+impl TryFrom<StrOrInt<'_>> for u32 {
+    type Error = ParseIntError;
+
+    fn try_from(v: StrOrInt) -> Result<Self, Self::Error> {
+        match v {
+            StrOrInt::Str(s) => s.parse(),
+            StrOrInt::Int(i) => Ok(i),
+        }
+    }
+}
+
+struct FilterVisitor<T>(PhantomData<T>);
+impl<'de> Visitor<'de> for FilterVisitor<u32> {
+    type Value = Filter<u32>;
+
+    fn expecting(&self, fmt: &mut Formatter) -> fmt::Result {
+        write!(fmt, "Expecting either a wildcard \"*\", decimal/hex value string, or list of decimal/hex value strings")
+    }
+
+    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        if v <= u32::MAX as u64 {
+            Ok(Self::Value::Enumerated(vec![v as u32]))
+        } else {
+            Err(E::custom("Id must fit within a u32 value"))
```

Comment:
Id?

Reply:
Domain Id, probably should say that.

Reply:
Oh, I had just used Id to make it generic, but since this code is only used here, it's fine.